File: python/gherkin/token_formatter_builder.py
from __future__ import annotations

from .ast_builder import AstBuilder
from .token import Token


class TokenFormatterBuilder(AstBuilder):
    _tokens: list[Token]

    def __init__(self) -> None:
        self.reset()
        self._tokens = []

    def reset(self) -> None:
        self._tokens = []

    def build(self, token: Token) -> None:
        self._tokens.append(token)

    def start_rule(self, rule_type: str) -> None:
        pass

    def end_rule(self, rule_type: str) -> None:
        pass

    def get_result(self) -> str:
        return "\n".join([self._format_token(token) for token in self._tokens])

    @staticmethod
    def _format_token(token: Token) -> str:
        if token.eof():
            return "EOF"

        return "".join(
            [
                "(",
                str(token.location["line"]),
                ":",
                str(token.location["column"]),
                ")",
                token.matched_type,
                ":",
                (
                    "".join(
                        [
                            "(",
                            (
                                token.matched_keyword_type
                                if token.matched_keyword_type
                                else ""
                            ),
                            ")",
                            token.matched_keyword,
                        ]
                    )
                    if token.matched_keyword
                    else ""
                ),
                "/",
                (token.matched_text if token.matched_text else ""),
                "/",
                ",".join(
                    [
                        str(item["column"]) + ":" + item["text"]
                        for item in token.matched_items
                    ]
                ),
            ]
        )

File: cpp/src/lib/gherkin/cucumber/gherkin/TokenFormatterBuilder.cpp
#include "cucumber/gherkin/TokenFormatterBuilder.hpp"
#include "cucumber/gherkin/IdGenerator.hpp"
#include "cucumber/gherkin/JoinUtils.hpp"
#include "cucumber/gherkin/RuleType.hpp"
#include "cucumber/gherkin/Token.hpp"
#include "cucumber/gherkin/Types.hpp"
#include <sstream>
#include <string>
#include <string_view>
#include <utility>

namespace cucumber::gherkin
{
    namespace
    {
        std::string FormatToken(const Token& token)
        {
            if (token.IsEof())
            {
                return "EOF";
            }

            std::ostringstream stream;

            stream << "(" << token.location.line << ":" << token.location.column.value_or(0) << ")" << token.matchedType << ":";

            if (token.matchedKeyword)
            {
                stream << "(";

                if (token.matchedKeywordType)
                {
                    stream << *token.matchedKeywordType;
                }

                stream << ")" << token.matchedKeyword.value();
            }

            stream << "/" << token.matchedText << "/";

            if (!token.matchedItems.empty())
            {
                Strings items;

                for (const auto& matchedItem : token.matchedItems)
                {
                    items.emplace_back(std::to_string(matchedItem.column) + ":" + matchedItem.text);
                }

                stream << Join(",", items);
            }

            return stream.str();
        }
    }

    TokenFormatterBuilder::TokenFormatterBuilder(IdGeneratorPtr idp)
        : idp(std::move(std::move(idp)))
    {}

    void TokenFormatterBuilder::Reset([[maybe_unused]] std::string_view uri)
    {
        formattedTokens.clear();
    }

    void TokenFormatterBuilder::StartRule(RuleType ruleType)
    {}

    void TokenFormatterBuilder::EndRule(RuleType ruleType)
    {}

    void TokenFormatterBuilder::Build(const Token& token)
    {
        formattedTokens.emplace_back(FormatToken(token));
    }

    Strings TokenFormatterBuilder::GetResult() const
    {
        return formattedTokens;
    }

}
